### PR TITLE
Issue 877: Add support for meters with the same name but different tag keys in prometheus registry

### PR DESCRIPTION
**Pull request by angry-cellophane**
I have an idea how to resolve this issue https://github.com/micrometer-metrics/micrometer/issues/877
and raising this PR to discuss it. 
The PR has a basic implementation + some simple tests to verify it works and makes sense in general.
I wanted to make sure that the solution makes sense first before going the full path and spending hours or days on making it bullet-proof and efficient. So any input is welcome.

So what this fix does:
1. Current version of micrometer-prometheus creates a collector for each pair of <metric name, list of tag keys>. Collector stores key values dynamically in a map.
2. Prometheus doesn't allow to register two collectors with the same metric name, but it doesn't limit the collector to one particular set of label names (or tag keys in micrometer). The main job of the collector is to return metric samples, which is a list of <tag key, tag value, value> basically. Prometheus supports metrics samples with the same metric name but different tag keys. 
3. So if the `MicrometerCollector` is changed to store tag keys dynamically like it does for tag values now, it can return tag keys and values in samples without any issue and prometheus will be able to parse it correctly.

What I did:
1. Added a new holder for tags to keep a list of keys and values separately for convenience.
2. Changed the map in `MicrometerCollector` to store the holders instead of tag values.
3. Changed some other dependent methods in `MicrometerCollector` to get tag keys from the holders.

### CODE REVIEW

**Comment by nhartner** [comment]
It would be great to see this merged. Is there any reason it hasn't been? Not sure who to tag on this but how about it @jonatan-ivanov?

**Comment by crolopez** [comment]
Any updates? The solution seems simple and would provide great value.

**Comment by jonatan-ivanov** [comment]
Please see this issue for more details: https://github.com/prometheus/client_java/issues/696
TL;DR: it possible but strongly discouraged.

**Comment by zorba128** [comment]
Please take a look at [/prometheus/client_java/issues/901].
While I might agree that having two metrics with different sets of labels is somehow problematic, I think this all is deeper problem that its disallowed to have same metrics published by two different collectors (same set of labels, just different label values).
And this is core problem that makes new api really hard to use anywhere but on simple code as on code snippets...

**Comment by tereigo** [comment]
Same problem here:
I've got a container-like app which can instantiate multiple instances of the business logic classes inside one jvm/process

Each instance exposes the same set of metrics

If I could set a label for each metric using the instance name and such metric objects are treated individually then it would work for this setup.

Otherwise I don't see any good solution apart from adding instance name into the metric name which is super ugly

**Comment by shakuzen** [comment]
@angry-cellophane thank you for this pull request. The code has shifted since this was opened because of the Prometheus Java client releasing a new major version and our Prometheus modules being split between the 0.x and 1.x clients. Finally in the recently released Prometheus Java client 1.5.0 the restriction in the client itself around metrics with the same name but a different set of label names has been lifted. In #7168 I applied changes similar to the ones here but a bit different. Let us know if there's any feedback. I adapted the MicrometerCollectorTest code you added here to the new Prometheus client API and committed it as 9188fe37be7ae0ed8c21da0cf68f865569458c2f. With that, I think we can close this PR and the linked issue is now resolved; the changes will be released in Micrometer 1.17.0-M3.
